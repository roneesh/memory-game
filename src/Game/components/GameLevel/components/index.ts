import LevelReady from './LevelReady';
// import LevelPlaying from './LevelPlaying';
// import LevelFinished from './LevelFinished';
// import LevelPreview from './LevelPreview';

export {
  LevelReady,
  // LevelPlaying,
  // LevelFinished,
  // LevelPreview,
};
