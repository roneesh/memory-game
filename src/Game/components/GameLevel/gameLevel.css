.game-board {
  background-color: #4f5d75;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-wrap: wrap;
  padding: 3px 0;
}

.game-board__tile {
  width: 100px;
  margin: 3px;
  height: 100px;
  color: #ffffff;
  background-color: #bfc0c0;
}