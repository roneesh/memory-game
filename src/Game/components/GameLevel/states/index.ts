// import LevelFinished from './LevelFinished';
// import LevelPlaying from './LevelPlaying';
import LevelPreview from './LevelPreview';
import LevelReady from './LevelReady';

export {
  // LevelFinished,
  // LevelPlaying,
  LevelPreview,
  LevelReady,
};
