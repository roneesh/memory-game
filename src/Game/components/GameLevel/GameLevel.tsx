import { shuffle } from 'lodash';
import * as React from 'react';
import './gameLevel.css';
import { LevelPreview, LevelReady, } from './states';

interface IGameLevelProps {
  updateLevel: () => void;
};

type LevelState = "ready" | "preview" | "playing" | "finished";
export interface IBoardTile {
  character: string | HTMLElement;
  visible: boolean;
};
export interface ILevelTurns {
  count: number;
  currentTurn: number[];
}
interface IGameLevelState {
  levelState: LevelState;
  board: IBoardTile[];
  turns:  ILevelTurns;
};

class GameLevel extends React.Component<IGameLevelProps, IGameLevelState> {
  constructor(props: IGameLevelProps) {
    super(props);

    this.state = {
      board: this.generateBoard(),
      levelState: 'ready',
      turns: {
        count: 0,
        currentTurn: [],
      },
    };
  }

  public generateBoard(): IBoardTile[] {
    return shuffle(
      ['A', 'A', 'B', 'B', 'C', 'C', 'D', 'D', 'E', 'E', 'F', 'F',
      'G', 'G', 'H', 'H', 'I', 'I', 'J', 'J', 'K', 'K', 'L', 'L'].map(character => {
        return {
          character,
          visible: false,
        };
      })
    );
  }

  public setLevelState = (levelState: LevelState) => {
    this.setState({ levelState });
  }

  public getResetBtn() {
    const { levelState } = this.state;

    const text = levelState === 'ready' ? 'Play Level' : 'Reset';
    const nextState = levelState === 'ready' ? 'preview' : 'ready';
    return <button onClick={this.setLevelState.bind(null, nextState)}>
      {text}
    </button>;
  }

  public renderControls() {

    return <div className="game-level__controls">
      {this.getResetBtn()}
    </div>
  }

  public renderState() {
    const { levelState, board } = this.state;
    switch(levelState) {
      case 'ready':
        return <LevelReady board={board} />;
      case 'preview':
        return <LevelPreview board={board} />;
      // case 'playing':
      //   return <LevelPlaying />;
      // case 'finished':
      //   return <LevelFinished />;
      default:
        return null;
    }
  }

  public render() {
    return <div className="game-level">
      {this.renderControls()}
      {this.renderState()}
    </div>;
  }
}

export default GameLevel;
