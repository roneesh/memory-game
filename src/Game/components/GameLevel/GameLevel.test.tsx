import { shallow } from 'enzyme';
import * as React from 'react';
import renderer from 'react-test-renderer';
import GameLevel from './GameLevel';

describe('The GameLevel', () => {

  const getComponent = (props?: any) => shallow(<GameLevel updateLevel={jest.fn()} {...props} />);
  const getSnapshot = (props?: any) => <GameLevel updateLevel={jest.fn()} {...props} />;

  it('it renders', () => {
    const gameHeader = renderer
      .create(getSnapshot())
      .toJSON();
    expect(gameHeader).toMatchSnapshot();
  });

  it('generates a board', () => {
    const component = getComponent();
    const board = (component.instance() as GameLevel).state.board;
    expect(board.length).toEqual(24);
  });

  describe('rendering different level states', () => {
    // it('renders a preview state', () => {
    //   const component = shallow(<GameLevel updateLevel={jest.fn()} />);
    //   component.setState({ state: 'preview'});
    //   expect(component.find(<LevelPreview />).length).toEqual(1);
    // });

    // it('renders a preview state', () => {
    //   const component = shallow(<GameLevel updateLevel={jest.fn()} />);
    //   component.setState({ state: 'playing'});
    //   expect(component.find(<LevelPlaying />).length).toEqual(1);
    // });

    // it('renders a finished state', () => {
    //   const component = shallow(<GameLevel updateLevel={jest.fn()} />);
    //   component.setState({ state: 'finished'});
    //   expect(component.find(<LevelFinished />).length).toEqual(1);
    // });
  });

  describe('rendering the level controls', () => {
    it('renders a button that starts the level', () => {
      const component = getComponent();
      const btn = component.find('button');
      btn.simulate('click');
      const instance = component.instance() as GameLevel;
      expect(instance.state.levelState).toEqual('preview');
    });

    it('renders a button to reset the level when playing', () => {
      const component = getComponent();
      const instance = component.instance() as GameLevel;
      instance.setLevelState('playing');
      const btn = component.find('button');
      btn.simulate('click');
      expect(instance.state.levelState).toEqual('ready');
    });
  });

});
