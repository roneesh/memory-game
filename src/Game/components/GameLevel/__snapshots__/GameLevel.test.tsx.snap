// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The GameLevel it renders 1`] = `
<div
  className="game-level"
>
  <div
    className="game-board"
  >
    <div
      className="game-board__tile"
    />
    <div
      className="game-board__tile"
    />
    <div
      className="game-board__tile"
    />
    <div
      className="game-board__tile"
    />
    <div
      className="game-board__tile"
    />
    <div
      className="game-board__tile"
    />
    <div
      className="game-board__tile"
    />
    <div
      className="game-board__tile"
    />
    <div
      className="game-board__tile"
    />
    <div
      className="game-board__tile"
    />
    <div
      className="game-board__tile"
    />
    <div
      className="game-board__tile"
    />
    <div
      className="game-board__tile"
    />
    <div
      className="game-board__tile"
    />
    <div
      className="game-board__tile"
    />
    <div
      className="game-board__tile"
    />
    <div
      className="game-board__tile"
    />
    <div
      className="game-board__tile"
    />
    <div
      className="game-board__tile"
    />
    <div
      className="game-board__tile"
    />
    <div
      className="game-board__tile"
    />
    <div
      className="game-board__tile"
    />
    <div
      className="game-board__tile"
    />
    <div
      className="game-board__tile"
    />
  </div>
  <div
    className="game-level__controls"
  >
    <button
      onClick={[Function]}
    >
      Play
    </button>
  </div>
</div>
`;
